fix: 404 not found when parent snippet is missing

enimaloc opened the pull request:
When the user duplicates an existing snippet, the server fetches the code
of the parent snippet via redis. When the snippet was missing, the server
was crashing due to a `KeyError("Snippet not found")`. This commit fixes
the problem by returning a `404 Snippet not found` HTTP error code and
status to the client in such case.

The problem is unlikely to happen under "normal" conditions as the snippet
must expires just between the moment the user fetches the snippet the
first time and the moment he clicks on the "duplicate" button.

Julien00859 commented on tests/test_controller.py:
```diff
@@ -142,6 +143,14 @@ def test_get_new_form(self):
+                with urlreq.urlopen("http://localhost:8012/?parentid=foo") as res:
+                    self.assertEqual(res.status, 404)
+                    self.html_sanitizer.feed(res.read().decode())
+
```
Ça devrait être dans un nouveau test `test_missing_parentid` par exemple

Julien00859 replied:
aussi je pense que le `with ... as res: self.assertEqual(res.status, 404)` n'est jamais exécuté à cause de l'exception, il faudait plutôt faire comme ceci:

```python
with self.assertRaises(HTTPError) as exc:
    urlreq.urlopen(...)

self.assertEqual(exc...)
```

Julien00859 reviewed (changes requested):
Merci pour ta contribution!

Quelques remarques sur le message de commit, ici ce n'est pas un ajout de fonctionnalité mais plutôt un correctif de bug. "handling" c'est un mot foure-tout qui au final n'a pas beaucoup de sens, je propose de ne retirer. "Actually" en anglais ça signifie "En fait", je pense que dans ce contexte-ci tu veux dire "Actuellement" le mot anglais pour ça est au choix "currently" ou "presently".

Je propose le wording suivant (à changer sur le message de commit directement, en plus du message sur la PR)

	fix: 404 not found when parent snippet is missing

	When the user duplicates an existing snippet, the server fetches the code
	of the parent snippet via redis. When the snippet was missing, the server
	was crashing due to a `KeyError("Snippet not found")`. This commit fixes
	the problem by returning a `404 Snippet not found` HTTP error code and
	status to the client in such case.

	The problem is unlikely to happen under "normal" conditions as the snippet
	must expires just between the moment the user fetches the snippet the
	first time and the moment he clicks on the "duplicate" button.

Alors attention je vois que tu as fait deux commits, j'aimerais qu'ils soient fusionnés avant qu'on merge. Pour fusionner des commits (et aussi profiter pour changer le message de commit) tu peux utiliser la commande rebase. La commande ci-dessous pour "rebaser" les 3 derniers commits, le message d'exemple est intéressant à lire, je pense que tu voudras utiliser une strategie `reword` pour le 1e commit et `fixup` pour le second.

    $ git rebase -i HEAD~4

Commit:
fix: 404 not found when parent snippet is missing

When the user duplicates an existing snippet, the server fetches the code
of the parent snippet via redis. When the snippet was missing, the server
was crashing due to a `KeyError("Snippet not found")`. This commit fixes
the problem by returning a `404 Snippet not found` HTTP error code and
status to the client in such case.

The problem is unlikely to happen under "normal" conditions as the snippet
must expires just between the moment the user fetches the snippet the
first time and the moment he clicks on the "duplicate" button.

Signed-off-by: enimaloc <github@enimaloc.ga>
